Heap/Heap.swift

```diff
@@ -89,69 +86,79 @@ public struct Heap<T> {
    * value (for a min-heap).
    */
   public func peek() -> T? {
-    return elements.first
+    return nodes.first
+  }
+  
+  /** Returns the node at given index */
+  public func node(at i: Int) -> T? {
+    guard i < nodes.count else { return nil }
+    return nodes[i]
   }
   
   /**
    * Adds a new value to the heap. This reorders the heap so that the max-heap
    * or min-heap property still holds. Performance: O(log n).
    */
   public mutating func insert(_ value: T) {
-    elements.append(value)
-    shiftUp(elements.count - 1)
+    nodes.append(value)
+    shiftUp(nodes.count - 1)
   }
   
+  /**
+   * Adds a sequence of values to the heap. This reorders the heap so that
+   * the max-heap or min-heap property still holds. Performance: O(log n).
+   */
   public mutating func insert<S: Sequence>(_ sequence: S) where S.Iterator.Element == T {
     for value in sequence {
       insert(value)
     }
   }
   
   /**
-   * Allows you to change an element. In a max-heap, the new element should be
-   * larger than the old one; in a min-heap it should be smaller.
+   * Allows you to change an element. This reorders the heap so that
+   * the max-heap or min-heap property still holds.
    */
   public mutating func replace(index i: Int, value: T) {
-    guard i < elements.count else { return }
+    guard i < nodes.count else { return }
     
-    assert(isOrderedBefore(value, elements[i]))
-    elements[i] = value
-    shiftUp(i)
+    removeAt(i)
+    insert(value)
   }
   
   /**
    * Removes the root node from the heap. For a max-heap, this is the maximum
    * value; for a min-heap it is the minimum value. Performance: O(log n).
    */
   @discardableResult public mutating func remove() -> T? {
-    if elements.isEmpty {
-      return nil
-    } else if elements.count == 1 {
-      return elements.removeLast()
-    } else {
-      // Use the last node to replace the first one, then fix the heap by
-      // shifting this new first node into its proper position.
-      let value = elements[0]
-      elements[0] = elements.removeLast()
-      shiftDown()
-      return value
+    if !nodes.isEmpty {
+      if nodes.count == 1 {
+        return nodes.removeLast()
+      } else {
+        // Use the last node to replace the first one, then fix the heap by
+        // shifting this new first node into its proper position.
+        let value = nodes[0]
+        nodes[0] = nodes.removeLast()
+        shiftDown(0)
+        return value
+      }
     }
+    return nil
   }
   
   /**
-   * Removes an arbitrary node from the heap. Performance: O(log n). You need
-   * to know the node's index, which may actually take O(n) steps to find.
+   * Removes an arbitrary node from the heap. Performance: O(log n).
+   * Note that you need to know the node's index.
    */
-  public mutating func removeAt(_ index: Int) -> T? {
-    guard index < elements.count else { return nil }
+  @discardableResult public mutating func removeAt(_ index: Int) -> T? {
```

Comment:
Let's update this signature to `remove(at index: Int)`.

Reply:
Thanks, for your further improvements really like them.